### PR TITLE
AZ-256: Add treasury pallet

### PR DESCRIPTION
**Pull request by pmikolajczyk41**
- `pallet_treasury` added; the parameters and setting were discussed offline
- the governance over treasury is passed to the same authority as `sudo`
- tests covering channeling fee and approving/rejecting proposals; the only thing missing is verifying whether tips are channeled as well; however, for now there is no easy way to do it - a new Jira ticket should be created unless discussed here otherwise

_explanation for tips problem_:
In out e2e tests we use `substrate-client-api` package. Unfortunately, it does not support specifying a tip for an extrinsic. There are basically three solutions, neither of which seems appropriate/possible for now:
1. extend `substrate-client-api` -- although we have already been contributing recently to their repository, we cannot integrate the newest commit, which operates on metadata v14 (giving up older v13 which we are using); hence, as long as we do not upgrade our code, we cannot enjoy new versions of  `substrate-client-api`
2. hard fork (probably temporarily) `substrate-client-api` and extend it with our needs -- this seems to be an overkill and introduces possible problems with maintanence
3. copy some functionality from `substrate-client-api` and edit it within our code -- very ugly
4. operate on hex data and do magic with appending tip -- awful

### CODE REVIEW

File: bin/runtime/src/lib.rs
Comment (timorl) [inline bin/runtime/src/lib.rs:285]: Maybe call this `EverythingToTheTreasury` or something, then you can avoid the comments below.

File: bin/runtime/src/lib.rs
Comment (timorl) [inline bin/runtime/src/lib.rs:456]: What does this ID do?

File: e2e-tests/src/main.rs
Comment (timorl) [inline e2e-tests/src/main.rs:121]: This would be better as a task in Jira.

File: e2e-tests/src/main.rs
Comment (timorl) [inline e2e-tests/src/main.rs:186]: This is a single function, but the above is two, confusing. Please make this consistent.

File: e2e-tests/src/utils.rs
Comment (timorl) [inline e2e-tests/src/utils.rs:121]: Yeah, this is why `util` modules are a bad idea.

Generally the organization of code around here is a bit of a mess, including a cluttered `main.rs`. It would be great if you could split the code somewhat more readably. Possible approach – split them by component tested and just import the test functions into main (making it tiny and keeping the per-component code in one place). I'm not sure how much work that would be though, if you think it's too much for this PR at least do this for the treasury tests.

You might still end up with a `util` module (e.g. containing the function above this comment?), but hopefully much smaller and if everything goes well with a better name (e.g. `finance` for the above?). Kinda guessing here, 'cause I'm not intimately familiar with this code.

If you have a better idea as to how to keep this organized that would likely also be fine.

File: e2e-tests/src/utils.rs
Comment (timorl) [inline e2e-tests/src/utils.rs:45]: These two functions here are kinda bizarre – you can just use `[0]` and `[1]` everywhere and this will panic when out of bounds (which is fine, the test will fail, and it's terribly unexpected if I understand correctly – it would mean the test has been set up incorrectly, right?).

File: bin/runtime/src/lib.rs
Comment (timorl) [reply bin/runtime/src/lib.rs:456]: Oh, it's the ID of the associated account? Why is it called what it's called then, is there a standard here?

File: e2e-tests/src/waiting.rs
Comment (timorl) [inline e2e-tests/src/waiting.rs:71]: One of these things is not like the others... is there no "approval" event?

**Review by timorl — CHANGES_REQUESTED** [review]
Some comments, mostly about code organization.

I'm slightly more worried that this changes the runtime and chainspec, and I don't feel confident that's a good idea right now, especially that we are working towards the testnet update which is turning out to be annoyingly complicated. I would feel safer if @DamianStraszak confirmed this is ok before merging (just the runtime+chainspec changes, not the whole PR).

File: bin/runtime/src/lib.rs
Comment (pmikolajczyk41) [reply bin/runtime/src/lib.rs:456]: Indeed, the `PalletId` is used to derive the corresponding account, an example in the pallet_treasury:
```rust
/// The treasury's pallet id, used for deriving its sovereign account ID.
#[pallet::constant]
type PalletId: Get<PalletId>;
```

In substrate these are some used names:
 - `PalletId(*b"py/trsry")`
 - `PalletId(*b"py/socie")`
 - `PalletId(*b"py/lotto")`

I followed their way. Also, afaik, it has to be of length 8

File: e2e-tests/src/main.rs
Comment (pmikolajczyk41) [reply e2e-tests/src/main.rs:121]: surely, but I wanted to wait for the review, in case somebody come up with a better solution to be applied here

File: e2e-tests/src/main.rs
Comment (pmikolajczyk41) [reply e2e-tests/src/main.rs:186]: I tried :P but unfortunately, these operations must be done in different ways :(
1. **approval** - sending an extrinsic is straightforward, but pallet_treasury does _not_ emit any corresponding event (apart from system.extrinsicSucceeded, but such event is in every block due to the timestamp extrinsic and there is no 'good' way of matching such result with the origin call); therefore, the only reliable way of verifying whether the approval was successful is just to look at the storage and check the `approvals` vec (and have a hope that the spending period has not just occurred :P)
2. **rejection** - again, sending an extrinsic is straightforward, and in this case there is a corresponding event, namely `Treasury::Rejected` so we can listen for this event; _however_, it often happens, that this event is dispatched before we start listening, so we need to setup listening before sending the extrinsic

File: e2e-tests/src/waiting.rs
Comment (pmikolajczyk41) [reply e2e-tests/src/waiting.rs:71]: answered in another conversation

File: bin/runtime/src/lib.rs
Comment (timorl) [reply bin/runtime/src/lib.rs:456]: > it has to be of length 8

...what? lol

In this case it cannot be improved much, but it would still be nice to know what the "py" refers to – it would be a bit silly if it was somehow supposed to be associated with e.g. Polkadot. Although the 'p' probably stands for "pallet"...?

File: e2e-tests/src/main.rs
Comment (timorl) [reply e2e-tests/src/main.rs:186]: Ah, these are definitely annoyances.

To make this less confusing I would suggest creating a `treasury_approve` function, probably just consisting of the two lines above, at least then they look the same in the test and the reader doesn't have to spend time wondering why you only check the approval. Internally the functions will still differ, but the code will be more readable.

File: bin/runtime/src/lib.rs
Comment (pmikolajczyk41) [reply bin/runtime/src/lib.rs:456]: regarding length: `pub struct PalletId(pub [u8; 8]);` (https://github.com/paritytech/substrate/blob/4581dd99ed59f65ede46a34c81133177d5824a68/frame/support/src/lib.rs#L111)

Comment (pmikolajczyk41) [reply bin/runtime/src/lib.rs:456]: image

Comment (pmikolajczyk41) [reply bin/runtime/src/lib.rs:456]: I dug it up, and now I wonder whether e.g. `a0/trsry` would be better :P

File: bin/runtime/src/lib.rs
Comment (DamianStraszak) [inline bin/runtime/src/lib.rs:115]: You also need to bump `transaction_version`

File: bin/runtime/src/lib.rs
Comment (timorl) [reply bin/runtime/src/lib.rs:456]: Excellent, I was afraid that it was something like this and well done finding such a direct proof. :D

"a0" as a prefix is indeed much better.

**Review by DamianStraszak — COMMENTED** [review]
When it comes to the runtime update associated with this PR:
1) I think it is fine to make changes in the chainspec, but it will not affect any already running chain, because according to my understanding we should always pass the old, original chainspec when running mainnet or testnet binaries. And even if we passed the new one then nothing would actually happen (??). But that is a topic worth researching...
2) I seems that the pallet does not have any storage items that are added in the genesis block (and I'm now pretty sure no pallet should ever do that -- and our does :P) so I think adding it without any migration should work with no issues. 
3) Even though 2), I think it would be great to merge this change after we make the first testnet update (hopefully Monday next week?), so that we don't update too many things at once, because, you know, we don't have much experience in updating yet.

File: bin/runtime/src/lib.rs
Comment (pmikolajczyk41) [reply bin/runtime/src/lib.rs]: renamed

File: bin/runtime/src/lib.rs
Comment (pmikolajczyk41) [reply bin/runtime/src/lib.rs]: renamed

File: e2e-tests/src/main.rs
Comment (pmikolajczyk41) [reply e2e-tests/src/main.rs]: Jira #329

File: e2e-tests/src/main.rs
Comment (pmikolajczyk41) [reply e2e-tests/src/main.rs]: done

File: e2e-tests/src/utils.rs
Comment (pmikolajczyk41) [reply e2e-tests/src/utils.rs]: I reorganized the package, but if it was just a transition between mess and mess', I can make further changes

File: e2e-tests/src/utils.rs
Comment (pmikolajczyk41) [reply e2e-tests/src/utils.rs]: done

File: bin/runtime/src/lib.rs
Comment (pmikolajczyk41) [reply bin/runtime/src/lib.rs]: bumped

File: e2e-tests/src/utils.rs
Comment (timorl) [inline e2e-tests/src/utils.rs:6]: This setup seems backwards to me. I would put all the `util` modules on the top level, and create a `test` module with all the modules that are currently in the top level. Then the test module would export all the functions that currently start with `test_*`, but without this prefix, and all main would do would be `use crate::test` and then call `test::fee_calculation` etc. There could also be a `lib.rs` file which would export all the utility functions from the currently-util modules and the things in `test` would import straight from `crate`.

This is incredibly pedantic tho, not a blocker for merging in the slightest, only do this if you are convinced it actually sounds better than the current version, and are in the mood. The current version is already massively better organized than what was here before, so thanks!

Comment (timorl) [reply e2e-tests/src/utils.rs:6]: Oh, and at least extrinsic.rs and types.rs would be merged into lib.rs in the above, maybe some others too (waiting.rs?).

**Review by timorl — APPROVED** [review]
Very nice, although keep Damian's comment in mind before merging.

One comment, but extremely nitpicky and feel free to ignore.

File: bin/runtime/src/lib.rs
Comment (fbielejec) [inline bin/runtime/src/lib.rs:451]: weird formatting?

File: bin/runtime/src/lib.rs
Comment (fbielejec) [inline bin/runtime/src/lib.rs:498]: same, weird formatting (but seems to be passing fmt checks :thinking: )

File: e2e-tests/src/main.rs
Comment (fbielejec) [inline e2e-tests/src/main.rs:22]: ince these test run single threaded, ooone thing comes to mind - we could use that fact to implicitely test finalization after changing validators (if there was ever to be a bug there), by moving the test earlier in the suite? For example:

```
    run(test::finalization, "finalization", config.clone())?;
    run(test::fee_calculation, "fee calculation", config.clone())?;
    run(test::change_validators, "validators change", config)?;
    run(test::token_transfer, "token transfer", config.clone())?;
    run(test::channeling_fee, "channeling fee", config.clone())?;
    run(test::treasury_access, "treasury access", config.clone())?;
```

File: bin/runtime/src/lib.rs
Comment (DamianStraszak) [reply bin/runtime/src/lib.rs:498]: It's because this is macro code. Fmt does not have power there :P

File: e2e-tests/src/test/treasury.rs
Comment (fbielejec) [inline e2e-tests/src/test/treasury.rs:108]: extra newline here?

File: e2e-tests/src/test/treasury.rs
Comment (fbielejec) [inline e2e-tests/src/test/treasury.rs:75]: put some real value pls

File: e2e-tests/src/test/treasury.rs
Comment (fbielejec) [inline e2e-tests/src/test/treasury.rs:107]: move to /utils/types.rs ?

File: e2e-tests/src/test/treasury.rs
Comment (fbielejec) [inline e2e-tests/src/test/treasury.rs:131]: to /utils/types.rs ?

**Review by fbielejec — COMMENTED** [review]
minor comment, solid work!

File: e2e-tests/src/test/treasury.rs
Comment (pmikolajczyk41) [reply e2e-tests/src/test/treasury.rs:131]: I have left this type (similarly to `ProposalTransaction`) as it is not returned from any public function and is for internal use only

File: e2e-tests/src/test/treasury.rs
Comment (pmikolajczyk41) [reply e2e-tests/src/test/treasury.rs:107]: similarly to `GovernanceTransaction`

File: e2e-tests/src/test/treasury.rs
Comment (pmikolajczyk41) [reply e2e-tests/src/test/treasury.rs]: changed

File: e2e-tests/src/main.rs
Comment (pmikolajczyk41) [reply e2e-tests/src/main.rs]: reordered

File: bin/runtime/src/lib.rs
Comment (pmikolajczyk41) [reply bin/runtime/src/lib.rs]: changed tabs to spaces (must have been introduced while copying from substrate :|)

File: bin/runtime/src/lib.rs
Comment (pmikolajczyk41) [reply bin/runtime/src/lib.rs]: changed tabs to spaces (must have been introduced while copying from substrate :|)

File: e2e-tests/src/utils.rs
Comment (pmikolajczyk41) [reply e2e-tests/src/utils.rs]: reorganized, thanks a lot!

**Review by fbielejec — APPROVED** [review]

**Review by timorl — APPROVED** [review]

**Pull request merged**
